src/planner/project_info.cpp

```diff
@@ -80,30 +82,30 @@ bool ProjectInfo::Evaluate(storage::Tuple *dest, const AbstractTuple *tuple1,
   return true;
 }
 
-bool ProjectInfo::Evaluate(AbstractTuple *dest, const AbstractTuple *tuple1,
-                           const AbstractTuple *tuple2,
-                           executor::ExecutorContext *econtext) const {
+bool ProjectInfo::Evaluate(expression::ContainerTuple<storage::TileGroup> *dest,
+                           expression::ContainerTuple<storage::TileGroup> *src,
+                           executor::ExecutorContext *econtext, bool inplace) const {
   // (A) Execute target list
   for (auto target : target_list_) {
     auto col_id = target.first;
     auto expr = target.second;
-    auto value = expr->Evaluate(tuple1, tuple2, econtext);
+    auto value = expr->Evaluate(src, nullptr, econtext);
     dest->SetValue(col_id, value);
   }
 
   // (B) Execute direct map
-  for (auto dm : direct_map_list_) {
-    auto dest_col_id = dm.first;
-    // whether left tuple or right tuple ?
-    auto tuple_index = dm.second.first;
-    auto src_col_id = dm.second.second;
-
-    if (tuple_index == 0) {
-      common::Value val1 = (tuple1->GetValue(src_col_id));
-      dest->SetValue(dest_col_id, val1);
-    } else {
-      common::Value val2 = (tuple2->GetValue(src_col_id));
-      dest->SetValue(dest_col_id, val2);
+  if (inplace == false) {
+    // Only copy un-modified value if not in-place update
+    for (auto dm : direct_map_list_) {
+      // whether left tuple or right tuple ?
+      auto tuple_index = dm.second.first;
+      auto src_col_id = dm.second.second;
+
+      PL_ASSERT(dm.first == dm.second.second);
+      PL_ASSERT(tuple_index == 0);
+      if (tuple_index == 0) {
+        src->CopyColumnTo(dest, src_col_id);
+      }
```

Comment:
I don't quite understand the logic here. The comment said "Only copy un-modified value if not in-place update", but it seems that the code suggested if "inplace == false", then we call CopyColumnTo(), which calls the shallow copy function. So should we do or not do a deep copy when it's in-place update? Also, shouldn't there be another branch of logic to do the deep copy on the other condition?